control/margins.py

```diff
@@ -54,26 +54,157 @@
+
+# private helper functions
+def _poly_iw(sys):
+    """Apply s = iw to G(s)=num(s)/den(s)
+
+    Splits the num and den polynomials with (iw) applied into a real and an
+    imaginary parts with w applied
+    """
+    num = sys.num[0][0]
+    den = sys.den[0][0]
+    num_iw = (1J)**np.arange(len(num) - 1, -1, -1) * num
+    den_iw = (1J)**np.arange(len(den) - 1, -1, -1) * den
+    return num_iw, den_iw
+
+
+def _poly_iw_sqr(pol_iw):
+    return np.real(np.polymul(pol_iw, pol_iw.conj()))
+
+
+def _poly_iw_real_crossing(num_iw, den_iw, epsw):
+    # Return w where imag(H(iw)) == 0
+    test_w = np.polysub(np.polymul(num_iw.imag, den_iw.real),
+                        np.polymul(num_iw.real, den_iw.imag))
+    w = np.roots(test_w)
+    w = np.real(w[np.isreal(w)])
+    w = w[w >= epsw]
+    return w
+
+
+def _poly_iw_mag1_crossing(num_iw, den_iw, epsw):
+    # Return w where |H(iw)| == 1, |num(iw)| - |den(iw)| == 0
+    w = np.roots(np.polysub(_poly_iw_sqr(num_iw), _poly_iw_sqr(den_iw)))
+    w = np.real(w[np.isreal(w)])
+    w = w[w > epsw]
+    return w
+
+
+def _poly_iw_wstab(num_iw, den_iw, epsw):
+    # Stability margin: minimum distance to point -1
+    # find zero derivative. Second derivative needs to be >0
+    # to have a minimum
+    test_wstabn = _poly_iw_sqr(np.polyadd(num_iw, den_iw))
+    test_wstabd = _poly_iw_sqr(den_iw)
+    test_wstab = np.polysub(
+        np.polymul(np.polyder(test_wstabn), test_wstabd),
+        np.polymul(np.polyder(test_wstabd), test_wstabn))
+
+    # find the solutions, for positive omega, and only real ones
+    wstab = np.roots(test_wstab)
+    wstab = np.real(wstab[np.isreal(wstab)])
+    wstab = wstab[wstab > epsw]
+
+    # and find the value of the 2nd derivative there, needs to be positive
+    wstabplus = np.polyval(np.polyder(test_wstab), wstab)
+    wstab = wstab[wstabplus > 0.]
+    return wstab
+
+
+def _poly_z_invz(sys):
+    num = sys.num[0][0]  # num(z) = a_p * z^p + a_(p-1) * z^(p-1) + ... + a_0
+    den = sys.den[0][0]  # num(z) = b_q * z^p + b_(q-1) * z^(q-1) + ... + b_0
+    p_q = len(num) - len(den)
+    if p_q > 0:
+        raise ValueError("Not a proper transfer function: Denominator must "
+                         "have equal or higher order than numerator.")
+    num_inv_zp = num[::-1]  # num(1/z) * z^p
+    den_inv_zq = den[::-1]  # den(1/z) * z^q
+    return num, den, num_inv_zp, den_inv_zq, p_q, sys.dt
+
+
+def _z_filter(z, dt, eps):
+    # z = exp(1J w dt)
+    # |z| == 1 with some float precision tolerance
+    z = z[np.abs(np.abs(z) - 1.) < eps]
+    zarg = np.angle(z)
+    zidx = (0 <= zarg) * (zarg < np.pi)
+    omega = zarg[zidx] / dt
+    return z[zidx], omega
+
+
+def _poly_z_real_crossing(num, den, num_inv_zp, den_inv_zq, p_q, dt, epsw):
+    # H(z)==H(1/z), num(z)*den(1/z) == num(1/z)*den(z)
+    p1 = np.polymul(num, den_inv_zq)
+    p2 = np.polymul(num_inv_zp, den)
+    if p_q < 0:
+        # Future: numpy >= 1.5.0 defines np.polymulx()
```

Comment:
Even NumPy itself doesn't support anything older than 1.16, do you really?

Reply:
Probably not

Reply:
Actually, polymulx is only in [np.polynomial.polynomial](https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.polymulx.html) which uses a different convention for the coefficients

```suggestion
        # * z**(-p_q)
```